fire call backs without lock in future

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30911 fire call backs without lock in future**
* #29579 build a generic future<T>

fire call backs without lock in future, follow up #29579

Differential Revision: [D18866475](https://our.internmc.facebook.com/intern/diff/D18866475/)

zhaojuanmao committed:
fire call backs without lock in future

fire call backs without lock in future, follow up #29579

Differential Revision: [D18866475](https://our.internmc.facebook.com/intern/diff/D18866475/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "fire call backs without lock in future"


fire call backs without lock in future, follow up #29579

Differential Revision: [D18866475](https://our.internmc.facebook.com/intern/diff/D18866475/)

[ghstack-poisoned]

pietern commented on torch/csrc/distributed/autograd/context/context.h:
```diff
@@ -112,7 +112,7 @@ class TORCH_API DistAutogradContext {
   // List of futures for RPCs initiated by this node to propagate gradients to
   // other nodes. The distributed autograd engine on this node can return
-  // successfully only if all these futures are done and are successful.
+  // successfully only if all these futures are done and are successfull.
```
This was correct before.

pietern commented on torch/csrc/utils/future.h:
```diff
@@ -47,18 +47,36 @@ class TORCH_API Future final {
 
-      fireCallbacks();
+      std::vector<Callback> cbs;
+      cbs.swap(callbacks_);
```
Can you add a comment why you're moving the callbacks here? This will be useful for the next person reading this code that doesn't know why this happens (e.g. "Move callbacks to a vector on the stack so we can access it without holding a lock").

pietern commented on torch/csrc/utils/future.h:
```diff
@@ -47,18 +47,36 @@ class TORCH_API Future final {
+      std::unique_lock<std::mutex> lock(mutex_);
+      TORCH_CHECK(!completed());
+      completed_ = true;
+      error_ = c10::guts::make_unique<FutureError>(std::move(errorMsg));
```
FYI, `c10::guts` can be changed to `std` very very shortly -- see #30915.

jjlilley commented on torch/csrc/utils/future.h:
```diff
@@ -47,18 +47,36 @@ class TORCH_API Future final {
+    {
+      std::unique_lock<std::mutex> lock(mutex_);
+      TORCH_CHECK(!completed());
+      completed_ = true;
```
minor: one other pattern that would be nice to fix here is setting completed_ before the value_ or error_.

I mostly raise this point because we allow completed() to access the atomic<bool> completed_ without holding the lock (which is ok).
This leads to the situation where completed() can momentarily be true, but the value_ is not actually set yet. I think the simple fix is to just set completed_ first
(though I guess this might not unequivocably work with memory reordering, but maybe closer)

zhaojuanmao replied:
@jjlilley are you suggesting to set completed_ after value_ and error_? because right now we've already set it first

jjlilley replied:
right, that's what I was suggesting.

But it might not matter in practice, the only reason I suggest doing so is because currently, completed() can be accessed without a lock, and can be true, without the variable actually being set.

zhaojuanmao committed:
Update on "fire call backs without lock in future"


fire call backs without lock in future, follow up #29579

Differential Revision: [D18866475](https://our.internmc.facebook.com/intern/diff/D18866475/)

[ghstack-poisoned]

zhaojuanmao commented:
merging this change to #29579, otherwise tests will fail in 29579 after rebasing. 

closing this
